File: backend/app/utils/key.py
def normalize_location_name(value) -> str:
    text = (
        str(value or "")
        .strip()
        .lower()
        .replace(".", "")
    )
    return "".join(ch for ch in text if ch.isalnum())


def build_location_key(district, mandal, village_name="") -> str:
    return "|".join(
        [
            normalize_location_name(district),
            normalize_location_name(mandal),
            normalize_location_name(village_name),
        ]
    )


def make_key(district, mandal, village_name="") -> str:
    return build_location_key(district, mandal, village_name)

File: frontend/src/utils/key.js
function normalizeKeyPart(value) {
  return String(value ?? "")
    .trim()
    .toLowerCase()
    .replace(/\./g, "")
    .replace(/\s+/g, " ")
    .replace(/[^a-z0-9]+/g, "");
}

export function buildDistrictVillageKey(district, mandal, villageName = "") {
  return [district, mandal, villageName].map(normalizeKeyPart).join("|");
}

export function buildMandalVillageKey(mandal, villageName = "") {
  return [mandal, villageName].map(normalizeKeyPart).join("|");
}

export const makeKey = buildDistrictVillageKey;

export function getDistrictVillageKeyFromRecord(record) {
  if (!record) return "";
  return buildDistrictVillageKey(
    record.district ?? record.District,
    record.mandal ?? record.Mandal,
    record.village_name ?? record.Village_Name ?? record.village ?? record.name
  );
}

export function normalizeDistrictVillageKeyPart(value) {
  return normalizeKeyPart(value);
}
